proposals/interface-types/working-notes/choices/webidl.md

```diff
@@ -0,0 +1,270 @@
+be used by an API designer that wishes to design for many different languages.
+
+There is a fine distinction to be made in making 'avoidance' recommendations and
+constructing a completely new IDL.
+
+#### WebIDL Strings
+
+Certain aspects of WebIDLs design seem to be somewhat overelaborated. For
+example, there are two variants of floating point numbers: float and
+unrestricted floats. (With corresponding equivalents for double precision
+floating point numbers.) The difference between a `float` and an `unrestricted
+float` in WebIDL is one that is not common in programming languages.
+
+Similarly,  WebIDL  has   three  variants  of  string   type:  the  `DOMString`,
+`ByteString` and `USVString` string. Again, the distinction between `DOMString`s
+and `USVString`s is not one that is common in programming languages.
+
+### Extended Attributes
+
+Heavy use is made in WebIDL of so-called extended attributes. For example, the
+`AllowShared` extended attribute is intended to mark a buffer type to be backed
+by a `SharedArrayBuffer`.
+
+Apart from the obvious dependence on a specific feature of JavaScript (shared
+array buffers), allowing sharing on values contradicts one of the key design
+criteria for Interface Types -- namely supporting interoperability in a cross
+ownership domain where there is no sharing.
+
+In practice, Web APIs that are specified using WebIDL make heavy use of extended
+attributes to convey additional semantics that are very JavaScript centric. Such
+usages would have to be prohibited when designing APIs for languages other than
+JavaScript.
+
+This again leads to a situation where the 'real' IDL that would be usable for
+WebAssembly would be a limited form of a public IDL. This is not an ideal model
+for standardization nor for supporting the wider WebAssembly ecosystem.
+
+### Nominal Types
+
+JavaScript does not have a static model of types. On the other hand, many
+programming languages do have static types. One of the features of many static type systems is the so-called _nominal type_.
+
+A nominal type differs from a structural type in several ways; but one of the
+most fundamental ways is that a nominal type can be used to model entities that
+are not directly conveyed by the actual data used to denote them.
+
+For example, one may choose to model a chair using the nominal type `Chair`;
+whose values contain a pair of integers: the SKU and price (for example). The
+pair of integers is _not equivalent_ to a chair, but represent the information
+needed to model chairs in some application or API. 
+
+A different entity (a person say) may also be modeled by a type `Person` which
+contains two integers (for example, an indentification number and an index into
+a street address table).
+
+A given pair of integers might be either a `Chair` or a `Person` (or a 2D point)
+and there is no way of conveying the distinction unless the nominal type is also
+taken into consideration.
```

Comment:
I don't follow here: WebIDL has the ability to capture nominal typing distinctions, like you've mentioned, via `interface`.

Reply:
Not really. Interface defines a set of operations. In a given system, neither Chair nor Person may have any special operations defined on them, they may (do) share the same data elements; yet they are still intended to denote different entities.

Reply:
I have the same reaction as Luke: WebIDL does have nominal types. I think the word `interface` is confusing here (like much of WebIDL's naming...), but in many places what's important about instances of DOM objects are their nominal type (which is checked at runtime when needed). Even JS has limited nominal types, like Array.

All of which is to say, I don't think this section is needed for your argument, and muddies the water by raising questions that, at least, don't land to some who have a lot of WebIDL knowledge.

Reply:
Agreed.

Reply:
WebIDL does not have nominal types. Interfaces are not the same thing.
However, I will delete.